3.13/engine.md

```diff
@@ -0,0 +1,270 @@
+on exits when they are cold. Since most exits will remain cold, we need to 
+store this information in a compact form.
+
+### Making "hot" exits fast and "cold" exits small
+
+In order to make hot exits fast, we them to be as simple as possible,
+passing as little information as possible.
+
+We also want to make cold exits as small as possible, but cold exits
+may become hot exits, so we need them both to use the same interface.
+
+## The implementation
+
+This section describes one possible implemnentation. The initial implementation
+will not support inter-superblock optimization, but we should plan to support it
+in the future.
+
+### Making "hot" exits fast
+
+In order to make hot exits fast, we want to implement them as a single, unconditional jump.
+In the tier 2 interpreter, that will be implemented by setting the IP to the first
+instruction of the target executor.
+In the JIT, it will be implemented as a tail call to the function pointer of the target executor.
+
+### Making "cold" exits small
+
+All side exits from an executor will start cold, and most of them will remain cold.
+
+We need to track various pieces of information for cold exits, none of which will be
+required once the exit becomes hot, so we want to store that information in a way
+that minimizes the cost to hot exits and minimizes the memory used for cold exits.
+That information is:
+* The offset/location of the pointer to the exit in the executor, so it can be updated.
+* The target (offset into the code object of the tier 1 instruction)
+* Any relevant known type information (this is optional but will improve optimization)
+* Any representation changes that have been made.
+* The "hotness" counter
+
+### Minimizing memory use
+
+There are number of ways we can reduce memory use, for example:
+* Putting things in (ideally const) arrays, so we can refer to them by a one or two byte index,
+  instead of an eight byte pointers.
+* For complex data like type information and representation changes, use trees or deltas,
+  so that the information like `(A, B, C)` can be stored as `(A, B) <- C` allowing `(A, B)` to
+  be shared.
+
+### Each executor gets a table of exit data
+
+Each executor will get a table of exit data. We can compute the size of this
+when creating the executor. The entries in this table are described below.
+
+### Fixed number of exit objects
+
+Since the vast majority of exits will not need patching, we do not want to pass the offset,
+so we need to store the offset in the executor.
+
+Since there is a fixed number of micro-ops allowed in a superblock, we have an uppper
+bound on the offset. We will-preallocate one exit object per possible offset.
+
+### Exit data
+
+* The offset/location of the pointer: Implicit if we embed the exit pointer in the exit data
+* The target: Has a maximum value of about 10**9, so store as a `uint32_t`
+* Any relevant known type information: Format TBD.
```

Comment:
Am I correct that this is only needed for intra-executor optimization and thus could be deferred until we start working on that?

Reply:
yes